Metrics documentation update (#171)
esl/MongoosePush: README.md

#### Available metrics

##### Histograms

For more details about the histogram metric type please go to https://prometheus.io/docs/concepts/metric_types/#histogram

###### Notification sent time

`mongoose_push_notification_send_time_microsecond_bucket{error_category=${CATEGORY},error_reason=${REASON},service=${SERVICE},status=${STATUS},le=${LE}}`
`mongoose_push_notification_send_time_microsecond_sum{error_category=${CATEGORY},error_reason=${REASON},service=${SERVICE},status=${STATUS}}`
`mongoose_push_notification_send_time_microsecond_count{error_category=${CATEGORY},error_reason=${REASON},service=${SERVICE},status=${STATUS}}`

Where:
* `STATUS` is `"success"` for the successful notifications or `"error"` in all other cases
* `SERVICE` is either `"apns"` or `"fcm"`
* `CATEGORY` is an arbitrary error category term (in case of `status="error"`) or an empty string (when `status="success"`)
* `REASON` is an arbitrary error reason term (in case of `status="error"`) or an empty string (when `status="success"`)
* `LE` defines the `less or equal` values for buckets, currently `1000`, `10_000`, `25_000`, `50_000`, `100_000`, `250_000`, `500_000`, `1000_000` or `+Inf`

> **NOTE**

> > A mesurment of value 50_000 will be added to all buckets which are less or equal to 50_000.

Reviewer: ```suggestion
> A measurement of value 50_000 will be added to all buckets which are less or equal to 50_000.
```

> > A mesurment of value 50_000 will be added to all buckets which are less or equal to 50_000.

Reviewer 2: Not sure if I understand this right, I think this is the other way around. The bucket contains measurements that are less or equal to the bucket size.

`More particularly they're counters which form a cumulative histogram, le stands for less than or equal to. So 26688 requests took less than or equal to 200ms, 27760 requests took less than or equal to 400ms, and there were 28860 requests in total. ` 
[https://www.robustperception.io/how-does-a-prometheus-histogram-work](https://www.robustperception.io/how-does-a-prometheus-histogram-work)

Author: Right, I messed with that. Good catch!

> In this case these are buckets `50_000`, `100_000`, `250_000`, `500_000`, `1000_000` and `+Inf`

This histogram metric shows the distribution of times needed to:
1. Select a worker (this may include waiting time when all workers are busy).
2. Send a request.
3. Get a response from push notifications provider.

##### Counters

> * `mongoose_push_supervisor_init_count{service=${SERVICE}}` - Counts number of push notification service supervisor starts.

Reviewer: ```suggestion
* `mongoose_push_supervisor_init_count{service=${SERVICE}}` - Counts the number of push notification service supervisor starts.
```

  The `SERVICE` variable can take `"apns"` or `"fcm"` as a value.

>   This metrics is update when MongoosePush starts or later when the underlying supervision tree is terminated and the error is propagate to the main application supervisor.

Reviewer: ```suggestion
  This metrics is updated when MongoosePush starts, and later on when the underlying supervision tree is terminated and the error is propagated to the main application supervisor.
```

>   This metrics is update when MongoosePush starts or later when the underlying supervision tree is terminated and the error is propagate to the main application supervisor.

Reviewer 3: ```suggestion
  This metric is updated when MongoosePush starts or later when the underlying supervision tree is terminated and the error is propagated to the main application supervisor.
```

> * `mongoose_push_apns_state_init_count` - Counts number of APNS state initialisations.

Reviewer: ```suggestion
* `mongoose_push_apns_state_init_count` - Counts the number of APNS state initialisations.
```

> * `mongoose_push_apns_state_terminate_count` - Counts number of APNS state terminations.

Reviewer: ```suggestion
* `mongoose_push_apns_state_terminate_count` - Counts the number of APNS state terminations.
```

> * `mongoose_push_apns_state_get_default_topic_count` - Counts number of default topic reads from cache.

Reviewer: ```suggestion
* `mongoose_push_apns_state_get_default_topic_count` - Counts the number of default topic reads from cache.
```

#### How to quickly see all metrics

```bash
curl -k https://127.0.0.1:8443/metrics
```

The above command assumes that MongoosePush runs on `localhost` and listens on port `8443`.

> Please, mind the `HTTPS` protocol, metrics are hosted on the same port as other API.

Reviewer: ```suggestion
Please, mind the `HTTPS` protocol, metrics are hosted on the same port than all the other API endpoints.
```
Or something like that, I'm not sure. I'd treat "API" as an indefinite singular noun, so "as other API" doesn't sound grammatically correct 🤔

> Please, mind the `HTTPS` protocol, metrics are hosted on the same port as other API.

Reviewer 2: ```suggestion
Please, mind the `HTTPS` protocol, metrics are hosted on the same port as other API endpoints.
```
How about this as compromise? 
`the same as` seems more grammatically correct than `the same than`

#### Prometheus configuration

When configuring Prometheus, it's important to:
* set the `scheme` to `https` since MongoosePush exposes `/metrics` path encrypted endpoint (HTTPS)
* set the `insecure_skip_verify` to `true` if the default self-signed certificates are used

```yaml
scrape_configs:
  - job_name: 'mongoose-push'
    scheme: 'https' #MongoosePush exposes encrypted endpoint - HTTPS
    tls_config: #The default certs used by MongoosePush are self-signed
      insecure_skip_verify: true #For checking purposes we can ignore certs verification
    static_configs:
      - targets: ['mongoose-push:8443']
        labels:
          group: 'production'

```

> 

Reviewer: Not needed empty lines
```suggestion
```
